### PR TITLE
Feature/build tools gulp

### PR DESCRIPTION
**Pull request by brianjhanson**
Sample of what it would look like with gulp as a build tool

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -33,20 +32,44 @@
     "not ie 11"
   ],
   "dependencies": {
-    "blendid": "^4.4.3",
-    "dotenv": "^6.0.0",
+    "gsap": "^3.4.2",
```
Comment (MikeMcChillin) [inline package.json:35]: How did this guy sneak in here?

File: gulpfile.js/tasks/images.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        svgoPlugins: [
+          {
+            cleanupIDs: false,
+            collapseGroups: false,
+            mergePaths: false,
+            moveElemsAttrsToGroup: false,
+            moveGroupAttrsToElems: false,
+            removeUselessStrokeAndFill: false,
+            removeViewBox: false,
+            removeStyleElement: true
```
Comment (MikeMcChillin) [inline gulpfile.js/tasks/images.js:23]: Wonder if we should move some of this to config? I could def see removing style elements becoming an issue.

File: src/scripts/app.js [hunk 3]
```diff
@@ -1,4 +1,13 @@
+import gsap from "gsap";
 import msg from "./modules/msg";
 
+// eslint-disable-next-line no-console
```
Comment (MikeMcChillin) [inline src/scripts/app.js:4]: if i need to put an eslint comment every time i `console.log`, i will die

File: package.json (hunk 1 above)
Comment (brianjhanson) [reply package.json:35]: This was just to test out the common chunks. I had to pull it in in order to get the bundle size up enough that it would split things out into separate bundles.

File: src/scripts/app.js (hunk 3 above)
Comment (brianjhanson) [reply src/scripts/app.js:4]: I mostly want to avoid `console.log` stuff making it to production, but I think we can update the lint rules so that it gives a warning locally, and gives an error when building.

The other option would be to write a simple function that we commit to always using to log, that's something like 
```js
function odcLog() {
  if (process.env.NODE_ENV === 'development') {
    console.log(arguments);
  }
}
```

**Comment by brianjhanson** [comment]
Just pushed two little commits:
1) Add the ability to specify postCss plugins within the config.js file. Right now, this would primarily be for tailwindCss so I have an example of that in here as well.

2) Moved the imageMin configuration out into the config.js file as well as per suggestion by @MikeMcChillin. This is a bit heavy handed, but it seemed like a better solution than something that would merge a deeply nested object in a smart way.

File: gulpfile.js/tasks/rev.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+const gulp = require("gulp");
+const rev = require("gulp-rev");
+const config = require("../config");
+
+/**
+ * Adds revision hash to assets and stores hashes in a manifest file
+ */
+function revAssets() {
+  return gulp
+    .src(`${config.destBase}/**/*.{css,js}`, { base: config.destBase })
```
Comment (MikeMcChillin) [inline gulpfile.js/tasks/rev.js:10]: Should this be updated to `css,js,jpg,png,woff2,woff`? Fonts & images are not being revved, but I think they should.

File: gulpfile.js/tasks/styles.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        extensions: [".scss"]
+      })
+    )
+    .pipe(
+      sass({
+        includePaths: ["node_modules"]
+      }).on("error", sass.logError)
+    )
+    .pipe(postcss(getPostCssProcessors(config.devMode)))
+    .pipe(sourcemaps.write())
```
Comment (MikeMcChillin) [inline gulpfile.js/tasks/styles.js:42]: Sourcemaps are being included in prod, which adds a lot of unnecessary weight.

File: .eslintrc.js [hunk 6]
```diff
@@ -4,6 +4,7 @@ module.exports = {
     "plugin:prettier/recommended"
   ],
   "rules": {
-    "func-names": ["warn", "never"]
+    "no-console": "off",
```
Comment (MikeMcChillin) [inline .eslintrc.js:7]: I'm getting errors for using `window` and `document`

Comment (MikeMcChillin) [reply .eslintrc.js:7]: This is an issue because libraries like alpine need access to the window, and I don't think we should force ourselves to write an eslint ignore line every time we want to write to the window.

File: package.json [hunk 7]
```diff
@@ -4,9 +4,8 @@
   "description": "Base craft install for ODC projects",
   "main": "index.js",
   "scripts": {
-    "start": "NODE_ENV=development blendid",
-    "build": "NODE_ENV=production blendid build",
-    "postinstall": "npm run build",
+    "start": "gulp",
```
Comment (MikeMcChillin) [inline package.json:7]: if I don't specifiy `NODE_ENV=development gulp`, I get an error that `NODE_ENV` is undefined.

**Comment by MikeMcChillin** [comment]
`PRIMARY_SITE_URL` and `NODE_ENV` are new required `.env` variables, so `.env.example` should be updated.

File: package.json (hunk 7 above)
Comment (brianjhanson) [reply package.json:7]: Hmm `NODE_ENV` should fall back to 'development'. Can you give more specifics around the error? We don't really want to set that in our .env I don't think

File: gulpfile.js/config.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    }
+  },
+  scripts: {
+    entry: {
+      app: [`${srcBase}/scripts/app.js`]
+    },
+    dest: `${destBase}/scripts`
+  },
+  templates: {
+    src: ["./templates/**/*.{html,twig}"]
```
Comment (MikeMcChillin) [inline gulpfile.js/config.js:38]: Should we take this opportunity to move everything into `app/`? Maybe you were planning on doing this already and it would just make this PR unreadable?

Comment (brianjhanson) [reply gulpfile.js/config.js:38]: I was going to ask the team about that but would likely do that in a separate PR because you're right, it would make this pretty hard to parse. I obviously like craft in the app folder, but didn't want to muck up too many workflows at once.

File: .eslintrc.js (hunk 6 above)
Comment (brianjhanson) [reply .eslintrc.js:7]: Just fixed this in 5b48ff7

File: gulpfile.js/config.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+          moveElemsAttrsToGroup: false,
+          moveGroupAttrsToElems: false,
+          removeUselessStrokeAndFill: false,
+          removeViewBox: false,
+          removeStyleElement: true
+        }
+      ]
+    }
+  },
+  watch: [
```
Comment (MikeMcChillin) [inline gulpfile.js/config.js:59]: Do we need a watch task for javascript? I noticed it's not livereloading properly when I add new `console.log`s to my `app.js` file.

File: package.json (hunk 7 above)
Comment (MikeMcChillin) [reply package.json:7]: Here's a screenshot of the `undefined` message i get in terminal when I don't include the `NODE_ENV` in the package.json. When I include it, it goes away.
Screen Shot 2020-08-14 at 2 00 52 PM

**Comment by chrismlusk** [comment]
Updates Node to close #48. Since this PR installs a lot of packages, I've added this here to reinstall everything with v12 and confirm all works as expected.

**Comment by chrismlusk** [comment]
@brianjhanson what do you think about splitting out the Webpack config from `scripts.js` and into `webpack.config.js` at the project root? It's a little funky because you end up importing the Gulp config from the `gulpfile.js` directory and into the Webpack config, and then you import the Webpack config back into the scripts task.

But I like it because:

- I think having a `webpack.config.js` file improves visibility about the setup.
- When my brain thinks "oh, I need to adjust Webpack," I can just search for "webpack" and get the right file.

**Comment by brianjhanson** [comment]
@chrismlusk yes, I put it in scripts because I was hoping to separate all the gulp tasks out into a separate package that we could pull in to past / non-craft projects a bit easier, but haven't gotten that far. 

With that in mind, I would propose we use something like [webpack merge](https://www.npmjs.com/package/webpack-merge) to merge a root `webpack.config` into the base config, at least eventually. 

For now, we can definitely just pull it into the root since I'm not sure if the separating things out will go as well as I hope (I think we can only easily provide a package for the tasks themselves, I'm not sure how easy it will be to provide the gulp-y part).

File: gulpfile.js/tasks/styles.js (hunk 5 above)
Comment (chrismlusk) [reply gulpfile.js/tasks/styles.js]: Added `gulp-if` to address this in c65faca

File: .nvmrc [hunk 10]
```diff
@@ -1 +1 @@
-v10.8.0
+v12.18.0
```
Comment (brianjhanson) [inline .nvmrc:1]: I think for the starter we should pin this to something less specific I kind of like `lts/*` but we could do `12.18` or even just `12` too.
```suggestion
lts/*
```

`lts/*` will pin this to whatever version of node is in long term stable, which is almost always what we want, especially since this build isn't super complex and we're relatively unlikely to run into version issues (as long as we keep the packages updated)

Comment (chrismlusk) [reply .nvmrc:1]: TIL!

File: gulpfile.js/tasks/rev.js (hunk 4 above)
Comment (brianjhanson) [reply gulpfile.js/tasks/rev.js:10]: The problem with revving fonts is that those are typically referenced by CSS which has no way of knowing what the revved paths are (unless we import them via Webpack or something)

File: package.json (hunk 7 above)
Comment (brianjhanson) [reply package.json:7]: This should be resolved by https://github.com/onedesign/craft-starter/pull/39/commits/fc5538b33a0a94073b5526802d0ec224b00b6e02 which explicitly sets the NODE_ENV is one is not set already
